hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -0,0 +1,383 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.Region
+import is.hail.asm4s.Code
+import is.hail.annotations._
+import is.hail.asm4s._
+import is.hail.asm4s.joinpoint._
+import is.hail.expr.ir.EmitMethodBuilder
+import is.hail.utils._
+
+object PCanonicalArray  {
+  def loadLength(aoff: Long): Int =
+    Region.loadInt(aoff)
+
+  def loadLength(aoff: Code[Long]): Code[Int] =
+    Region.loadInt(aoff)
+
+  def storeLength(aoff: Long, length: Int): Unit =
+    Region.storeInt(aoff, length)
+
+  def storeLength(aoff: Code[Long], length: Code[Int]): Code[Unit] =
+    Region.storeInt(aoff, length)
+
+  def nMissingBytes(len: Code[Int]): Code[Int] = (len + 7) >>> 3
+
+  def nMissingBytes(len: Int): Long = (len + 7L) >>> 3
+}
+
+final case class PCanonicalArray(elementType: PType, required: Boolean = false) extends PArray {
+  def _asIdent = s"array_of_${elementType.asIdent}"
```

Comment:
Left these as array_of, Array,array, because I think it's reasonable to call the canonical implementation the same as the abstract type. We tend to be ok with not padding numbers with 0 for instance.